api/src/test/java/org/pmiops/workbench/cdr/dao/WorkspaceDaoTest.java

```diff
@@ -0,0 +1,49 @@
+package org.pmiops.workbench.cdr.dao;
+
+import static org.springframework.test.util.AssertionErrors.fail;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.pmiops.workbench.db.dao.WorkspaceDao;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.context.annotation.Import;
+import org.springframework.orm.ObjectOptimisticLockingFailureException;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.annotation.DirtiesContext.ClassMode;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+@Import(LiquibaseAutoConfiguration.class)
+@AutoConfigureTestDatabase(replace= AutoConfigureTestDatabase.Replace.NONE)
+@DirtiesContext(classMode = ClassMode.BEFORE_EACH_TEST_METHOD)
+@Transactional(propagation = Propagation.NOT_SUPPORTED)
+public class WorkspaceDaoTest {
+  @Autowired
+  WorkspaceDao workspaceDao;
+
+  @Test
+  public void testWorkspaceVersionLocking() {
+    org.pmiops.workbench.db.model.Workspace ws = new org.pmiops.workbench.db.model.Workspace();
+    ws.setVersion(1);
+    ws = workspaceDao.save(ws);
+
+    // Version incremented to 2.
+    ws.setName("foo");
+    ws = workspaceDao.save(ws);
+
+    try {
+      ws.setName("bar");
+      ws.setVersion(1);
+      workspaceDao.save(ws);
+      fail("expected optimistic lock exception on stale version update");
+    } catch(ObjectOptimisticLockingFailureException e) {
+      // expected
```

Comment:
Is there a reason not to make the test `@Test(expected= ObjectOptimisticLockingFailureException.class)`?

Reply:
I wanted to explicitly assert which callsite yielded the exception.